cmd/katapult/vms.go

```diff
@@ -0,0 +1,230 @@
+package main
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/krystal/go-katapult"
+	"github.com/krystal/go-katapult/core"
+	"github.com/spf13/cobra"
+)
+
+type virtualMachinesClient interface {
+	List(
+		ctx context.Context,
+		org core.OrganizationRef,
+		opts *core.ListOptions,
+	) ([]*core.VirtualMachine, *katapult.Response, error)
+
+	Delete(
+		ctx context.Context,
+		ref core.VirtualMachineRef,
+	) (*core.TrashObject, *katapult.Response, error)
+
+	Shutdown(
+		ctx context.Context,
+		ref core.VirtualMachineRef,
+	) (*core.Task, *katapult.Response, error)
+
+	Start(
+		ctx context.Context,
+		ref core.VirtualMachineRef,
+	) (*core.Task, *katapult.Response, error)
+
+	Stop(
+		ctx context.Context,
+		ref core.VirtualMachineRef,
+	) (*core.Task, *katapult.Response, error)
+
+	Reset(
+		ctx context.Context,
+		ref core.VirtualMachineRef,
+	) (*core.Task, *katapult.Response, error)
+}
+
+func vmFlag(cmd *cobra.Command) (core.VirtualMachineRef, error) {
+	id := cmd.Flag("id").Value.String()
+	ref := core.VirtualMachineRef{ID: id}
+	if id == "" {
+		fqdn := cmd.Flag("fqdn").Value.String()
+		if fqdn == "" {
+			return core.VirtualMachineRef{}, fmt.Errorf("both ID and FQDN are unset")
+		}
+		return core.VirtualMachineRef{FQDN: fqdn}, nil
+	}
+	return ref, nil
+}
+
+func vmNotFoundHandlingError(err error, resp *katapult.Response) error {
+	if resp != nil && resp.StatusCode == 404 {
+		return fmt.Errorf("unknown virtual machine")
+	}
+	return err
+}
+
+func virtualMachinesListCmd(client virtualMachinesClient) *cobra.Command {
+	list := &cobra.Command{
+		Use:     "list",
+		Aliases: []string{"ls"},
+		Short:   "Get a list of virtual machines from an organization",
+		Long:    "Get a list of virtual machines from an organization.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			ref := core.OrganizationRef{ID: cmd.Flags().Lookup("id").Value.String()}
+			if ref.ID == "" {
+				subdomain := cmd.Flags().Lookup("subdomain").Value.String()
+				if subdomain == "" {
+					return fmt.Errorf("both ID and subdomain are unset")
+				}
+				ref.SubDomain = subdomain
+			}
+
+			stdout := cmd.OutOrStdout()
+
+			page := 1
+			perPage := 30
+			for {
+				vms, resp, err := client.List(cmd.Context(), ref, &core.ListOptions{
+					Page:    page,
+					PerPage: perPage,
+				})
+				if err != nil {
+					return err
+				}
+
+				// Ensure all the data for the pagination is correct.
+				pagination := resp.Pagination
+				page = pagination.CurrentPage + 1
+				perPage = pagination.PerPage
+				end := pagination.TotalPages == 0 || pagination.CurrentPage == pagination.TotalPages
+
+				for _, vm := range vms {
+					fqdn := vm.FQDN
+					if fqdn == "" {
+						fqdn = "<no fqdn specified>"
+					}
```

Comment:
I don't think this check is needed. As far as I know, VMs will always have a FDQN value.